### DISCUSSION TITLE
Serializing Container of Base Class Objects

### DISCUSSION THREAD

**Discussion by jmarcelo2397** [nlohmann/json#2935] [Q&A]
Let's say that I have a set of classes derived from a base class:

    class Base {};
    void to_json(json& j, const Base& obj);
    
    class Derived1 : public Base {}
    void to_json(json& j, const Derived1& obj); // calls "update()" with additional members
    
    class Derived2 : public Base {}
    void to_json(json& j, const Derived2& obj); // calls "update()" with additional members

...and I have an STL container of the base classes:

    std::vector<Base> objects;

If I serialize & dump the vector, the objects are only interpreted as their base class & do not include the additional members from the derived classes. How can I serialize those derived classes in the container?

**Comment by nlohmann** [comment]
Maybe related: #2223 #2199 #803

**Reply by jmarcelo2397** [reply]
Yes I have referenced all of those but they do not address my issue. Let me clarify.

**What do you want to achieve?**

I want to serialize a container of objects that are all derived from the same base class.

**What have you tried?**

I've written the `to_json()` function for the base class, and used `update()` function on the derived classes to add additional members. But, when putting the items into a container of their base classes & serializing the container, each item loses their derived data members.

**Can you provide a small code example?**

Not small, but to illustrate my point:

Let's say we have the following base class:

    class Shape {
      public:
        string _name;
        Shape(string name) : _name(name){}
    };

...now let's derive some subclasses:

    class Circle : public Shape {
      public:
        unsigned _radius;
        Circle(string name, unsigned radius): Shape(name), _radius(radius) {}
    };
    
    class Square : public Shape {
      public:
        unsigned _length;
        Square(string name, unsigned length): Shape(name), _length(length) {}
    };

Now let's write the `to_json()` functions for each:

    void to_json(json& j, const Shape& s)
    {
        j = {{ "_name", s._name }};
    }
    
    void to_json(json&j, const Circle& c)
    {
        to_json(j, static_cast<Shape>(c));
        j.update({{ "_radius", c._radius }});
    }
    
    void to_json(json&j, const Square& s)
    {
        to_json(j, static_cast<Shape>(s));
        j.update({{ "_length", s._length }});
    }

Then, let's create a `vector` of type `Shape` and add a `Circle` & `Square`:

      Circle c("circle", 5);
      Square s("square", 10);
      vector<Shape> v{ c, s };

...and finally print all 3:

      json j1 = c;
      cout << j1.dump(4) << endl;
  
      json j2 = c;
      cout << j2.dump(4) << endl;
  
      json j3 = v;
      cout << j3.dump(4) << endl;

**Expected Output**

    {
        "_name": "circle",
        "_radius": 5
    }
    {
        "_length": 10,
        "_name": "square"
    }
    [
        {
            "_name": "circle",
            "_radius": 5
        },
        {
            "_length": 10,
            "_name": "square"
        }
    ]

**Actual Output**

    {
        "_name": "circle",
        "_radius": 5
    }
    {
        "_length": 10,
        "_name": "square"
    }
    [
        {
            "_name": "circle" // no _radius
        },
        {
            "_name": "square" // no _length
        }
    ]

**Reply by gregmarr** [reply]
This issue has nothing to do with the library.  Your data loss is here:

    Circle c("circle", 5);
    Square s("square", 10);
    vector<Shape> v{ c, s };

You now have a vector of objects of type `Shape` containing just the `Shape` portions of `c` and `s`.  If you want to be able to put derived objects into the vector, you need to use `vector<Shape *> v;` and put pointers in there, and then you have to manage the memory for those objects.

**Reply by jmarcelo2397** [reply]
Related: #2377 

Okay - so I should use pointers to `Shape` rather than `Shape`. 2 follow up questions:

1. What do you mean by:
> manage the memory for those objects.

2. Even if I have `vector<Shape*> v` or actually, `vector<shared_ptr<Shape> >`, when I serialize it into json & dump, it still only reads the objects as their base class `Shape` rather than their respective derived classes. How would I go about treating each object as its derived class?

**Reply by jmarcelo2397** [reply]
Actually, the best solution I've found so far is to have a member in the base class that allows us to differentiate different derived class types:

    typedef enum {
        SQUARE,
        CIRCLE
    } category_t;

    class Shape {
      public:
        string _name;
        category_t _category;
        Shape(string name, category_t category) : _name(name), _category(category){}
    };

...then write the following serialization function:

    void to_json(json& j, const vector<shared_ptr<Shape> >& v)
    {
        for (const auto& shape: v)
        {
            if (shape->_category == CIRCLE)
            {
                j.push_back(*(static_pointer_cast<Circle>(shape)));
            }
            else {
                j.push_back(*(static_pointer_cast<Square>(shape)));
            }
        }
    }

**Comment by gregmarr** [comment]
> manage the memory for those objects.

You covered that with this:

> Even if I have vector<Shape*> v or actually, vector<shared_ptr<Shape> >

>Actually, the best solution I've found so far is to have a member in the base class that allows us to differentiate different derived class types:

If you don't want to add a data member, you can use `std::dynamic_pointer_cast<T>` instead, or even `dynamic_cast<T>`.

    for (const auto& shape: v)
    {
        if (auto circle = dynamic_pointer_cast<Circle>(shape))
        {
            j.push_back(*circle);
        }
        else if (auto square = dynamic_pointer_cast<Square>(shape)) 
        {
            j.push_back(*square);
        }
    }


    for (const auto& shape: v)
    {
        if (auto circle = dynamic_cast<Circle*>(shape.get()))
        {
            j.push_back(*circle);
        }
        else if (auto square = dynamic_cast<Square*>(shape.get())) 
        {
            j.push_back(*square);
        }
    }

Update: fixed dynamic_cast to include the missing `*`.
